File: answers/lintcode/Combinations.cpp
class Solution {
public:
    /**
     * @param n: Given the range of numbers
     * @param k: Given the numbers of combinations
     * @return: All the combinations of k numbers out of 1..n
     */
    vector<vector<int> > combine(int n, int k) {
        // write your code here
		vector<vector<int> > comb_list;
		vector<int> comb;
		combine(1, comb, n, k, comb_list);
		return comb_list;
    }
	void combine(int current, vector<int> &comb, int n, int k, vector<vector<int> > &comb_list) {
		if (comb.size() >= k) {
			comb_list.push_back(comb);
			return;
		}
		for (int i = current; i <= n; ++ i) {
			comb.push_back(i);
			combine(i + 1, comb, n, k, comb_list);
			comb.pop_back();
		}
	}
};

File: answers/leetcode/Combinations.py
class Solution(object):
    def combine(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: List[List[int]]
        """

        combinations = []
        self.dfs(1, 0, [], combinations, n, k)
        return combinations

    def dfs(self, begin, count, combi, combinations, n, k):
        if count >= k:
            combinations.append(combi)
            return
        for i in range(begin, n + 1):
            self.dfs(i + 1, count + 1, combi + [i], combinations, n, k)
